omod/src/main/java/org/openmrs/module/fhir2/web/filter/SmartAuthenticationFilter.java

```diff
@@ -65,29 +67,56 @@ public void init(FilterConfig filterConfig) {
 	public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain) throws IOException, ServletException {
 		if (req instanceof HttpServletRequest) {
 			final HttpServletRequest httpRequest = (HttpServletRequest) req;
-			
-			// KeycloakOIDCFilter does the actual request handling
-			super.doFilter(req, res, (rq, rs) -> {});
-			
 			if (httpRequest.getRequestedSessionId() != null && !httpRequest.isRequestedSessionIdValid()) {
 				Context.logout();
 			}
 			
-			if (httpRequest.getAttribute(KeycloakAccount.class.getName()) != null) {
-				KeycloakAccount account = (KeycloakAccount) httpRequest.getAttribute(KeycloakAccount.class.getName());
+			if (!Context.isAuthenticated()) {
+				// KeycloakOIDCFilter does the actual request handling
+				super.doFilter(req, res, (rq, rs) -> {});
 				
-				User user = new User();
-				user.setUsername(account.getPrincipal().getName());
-				
-				Authenticated authenticated = Context.authenticate(new SmartTokenCredentials(user));
+				if (httpRequest.getRequestedSessionId() != null && !httpRequest.isRequestedSessionIdValid()) {
+					Context.logout();
+				}
 				
-				log.debug("The user '{}' was successfully authenticated as OpenMRS user {}",
-				    account.getPrincipal().getName(), authenticated.getUser());
-			} else {
-				if (!res.isCommitted()) {
-					HttpServletResponse httpResponse = (HttpServletResponse) res;
-					httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Not authenticated");
-					return;
+				if (httpRequest.getAttribute(KeycloakAccount.class.getName()) != null) {
+					OidcKeycloakAccount account = (OidcKeycloakAccount) httpRequest
+					        .getAttribute(KeycloakAccount.class.getName());
+					
+					String userName = account.getKeycloakSecurityContext().getToken().getSubject();
+					Authenticated authenticated = Context.authenticate(new SmartTokenCredentials(userName));
+					
+					log.debug("The user '{}' was successfully authenticated as OpenMRS user {}", userName,
+					    authenticated.getUser());
+					
+					// HAPI cannot process the "access_token" parameter, so we remove it and forward the request
+					if (httpRequest.getParameter("access_token") != null && httpRequest.getQueryString() != null) {
+						String queryString = httpRequest.getQueryString();
+						String newQueryString = ACCESS_TOKEN.matcher(queryString).replaceFirst("");
+						
+						StringBuilder newRequestUri = new StringBuilder();
+						if (httpRequest.getContextPath() != null) {
+							newRequestUri.append(httpRequest.getContextPath());
+						}
+						
+						if (httpRequest.getServletPath() != null) {
+							newRequestUri.append(httpRequest.getServletPath());
+						}
+						
+						if (httpRequest.getPathInfo() != null) {
+							newRequestUri.append(httpRequest.getPathInfo());
+						}
+						
+						newRequestUri.append('?').append(newQueryString);
+						
+						req.getRequestDispatcher(newRequestUri.toString()).forward(req, res);
+					}
```

Comment:
@ibacher Don't you think that we still need the `ForwardingFilter` to execute?

Also the there is a mistake in the url formation it adds /openmrs/openmrs two times

Reply:
This isn't intended to override the ForwardingFilter, only to remove the `access_token` from the query before it hits HAPI. Note that this "forwarding" is really just the way to internally change the URL without having to worry about too many details.

> Also there is a mistake in the url formation it adds /openmrs/openmrs two times

That clearly needs to be fixed.